### ISSUE TITLE
Set execution time of a specific query

### ISSUE THREAD

**Issue by ibazulic** [coleifer/peewee#2926]
Hey there,

Is there a way to limit the time an SQL execution takes place? For instance, doing something like:

~~~
...
try:
  Users.select().limit(1).timeout(5)
except TimeOutReached as ex:
  logger.debug("Exception raised: %s", ex)
...
~~~

I haven't found anything in the documentation, but I might have missed it. I could program a solution that would raise an exception after a certain time has passed, but I'm not sure in what condition will that leave the request that is being executed, i.e. will I just interrupt normal program flow or will the interruption actually kill Peewee's attempt to contact the database and abandon the request completely.

Any hints? Thanks!

**Comment by coleifer** [comment]
That would be nice, but that's not provided for in the [DB-API 2.0](https://peps.python.org/pep-0249/) and database implementation is sketchy.

For postgres (and possibly mysql) you can use a context manager:

```
@contextmanager
def psql_timeout(db, s):
    db.execute_sql('set statement_timeout=%s', (s,))
    try:
        yield db
    finally:
        db.execute_sql('set statement_timeout=0')

with psql_timeout(db, 5000):
    # queries here, each query capped at 5s.
```

For sqlite you probably need to implement a timer that calls `interrupt` on the connection after the time elapses - you can call back from C to python using a progress handler like here: https://github.com/simonw/datasette/blob/2170269258d1de38f4e518aa3e55e6b3ed202841/datasette/utils/__init__.py#L187 or use a separate thread although sharing connections between threads can be weird w/python.

In short I'd suggest adapting whatever approach matches the db you are using.

**Comment by ibazulic** [comment]
Thanks @coleifer for providing such a quick answer! We are using PostgreSQL so your example most likely will work for us. Have a great day!
